### ISSUE TITLE
Confusing `owned_box` error message since nightly-2019-09-04

### ISSUE THREAD

**Issue by phil-opp** [rust-lang/rust#64430]
Test Program:

```rust
// lib.rs

#![no_std]

pub fn foo() {
    pub struct Writer;
    write!(Writer, "");
}
```

Error message on nightly [_9af17757b 2019-09-02_] and before:

```
error[E0599]: no method named `write_fmt` found for type `foo::Writer` in the current scope
 --> src/lib.rs:5:5
  |
4 |     pub struct Writer;
  |     ------------------ method `write_fmt` not found for this
5 |     write!(Writer, "");
  |     ^^^^^^^^^^^^^^^^^^^
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `write_fmt`, perhaps you need to implement it:
          candidate #1: `core::fmt::Write`
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Error message since nightly [_b9de4ef89 2019-09-03_]:

```
error: requires `owned_box` lang_item

error: aborting due to previous error
```

**Diff Between the Nightlies:** https://github.com/rust-lang/rust/compare/9af17757be1cc3f672928ecf06c40a662c5ec26d...b9de4ef89e0e53099a084001b26ec3207c5f8391

---

Reported by @wgfm and @ambye85 in https://github.com/phil-opp/blog_os/issues/405#issuecomment-530537024.

**Comment by 12101111** [comment]
I'm testing this with [cargo-bisect-rustc](https://github.com/rust-lang-nursery/cargo-bisect-rustc/)

**Comment by Centril** [comment]
cc @estebank -- I think we need to be more graceful by checking if liballoc exists or not or not requiring the language items.

**Comment by 12101111** [comment]
PR #64056 fix #62373, show more suggestions on method take `Box<Self>` ,`Pin<Self>`, `Arc<Self>` or `Rc<Self>` when method not found
```rust
        if let Some(mut err) = self.report_method_error(
            span,
            rcvr_t,
            segment.ident,
            SelfSource::MethodCall(rcvr),
            error,
            Some(args),
        ) {
            if let ty::Adt(..) = rcvr_t.sty {
                // Try alternative arbitrary self types that could fulfill this call.
                // FIXME: probe for all types that *could* be arbitrary self-types, not
                // just this whitelist.
                let box_rcvr_t = self.tcx.mk_box(rcvr_t);
                try_alt_rcvr(&mut err, box_rcvr_t);
                let pin_rcvr_t = self.tcx.mk_lang_item(
                    rcvr_t,
                    lang_items::PinTypeLangItem,
                );
                try_alt_rcvr(&mut err, pin_rcvr_t);
                let arc_rcvr_t = self.tcx.mk_lang_item(rcvr_t, lang_items::Arc);
                try_alt_rcvr(&mut err, arc_rcvr_t);
                let rc_rcvr_t = self.tcx.mk_lang_item(rcvr_t, lang_items::Rc);
                try_alt_rcvr(&mut err, rc_rcvr_t);
            }
            err.emit();
        }
```
However,`Box` , `Arc` and `Rc` don't exist in `core`, but in `alloc`.

```rust
#![no_std]
extern crate alloc;

fn main() {
    pub struct Writer;
    write!(Writer, "");
}
```
Error message of this code: 
```log
error: no global memory allocator found but one is required; link to std or add `#[global_allocator]` to a static item that implements the GlobalAlloc trait.

error: `#[panic_handler]` function required, but not found

error: language item required, but not found: `eh_personality`

error: `#[alloc_error_handler]` function required, but not found

error[E0599]: no method named `write_fmt` found for type `main::Writer` in the current scope
 --> src/main.rs:6:5
  |
5 |     pub struct Writer;
  |     ------------------ method `write_fmt` not found for this
6 |     write!(Writer, "");
  |     ^^^^^^^^^^^^^^^^^^^ method not found in `main::Writer`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `write_fmt`, perhaps you need to implement it:
          candidate #1: `core::fmt::Write`
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 5 previous errors
```

**Comment by Centril** [comment]
Mentoring instructions:

- Change `mk_lang_item` to return `Option<Ty<'tcx>>` and use `self.lang_items().require(lang_item)` instead of `require_lang_item`.

- Make `try_alt_rcvr` take in `recv_t` and call `mk_lang_item` instead and return early `if let None = new_rcvr_t`. Also be sure to not use `mk_box`.

- Include tests like above and make sure to check for Box, Rc, and Arc.

- Use `./x.py -i test src/test/ui --stage 1 --bless --pass check` to update the tests.
